File: Medium/Bulls and Cows/bulls_and_cows.py
# Date: 25/05/22

class Solution:

    def getHint(self, secret: str, guess: str) -> str:
        secret_dict = self.createDictionary(secret)
        guess_dict = self.createDictionary(guess)
        cows = 0
        bulls = 0
        for i in range(len(secret)):
            if secret[i] == guess[i]:
                bulls += 1
                secret_dict[secret[i]] -= 1
                guess_dict[guess[i]] -= 1
        for i in secret:
            if secret_dict[i] > 0 and i in guess_dict and guess_dict[i] > 0:
                cows += 1
                secret_dict[i] -= 1
                guess_dict[i] -= 1
        return str(bulls)+'A'+str(cows)+'B'

    def createDictionary(self, string: str) -> dict:
        new_dictionary = {}
        for c in string:
            new_dictionary.setdefault(c, 0)
            new_dictionary[c] += 1
        return new_dictionary

File: Medium/Bulls and Cows/bulls_and_cows.rs
// Date: 25/05/22
// Runtime: 6 ms, faster than 7.14% of Rust online submissions for Bulls and Cows.
// Memory Usage: 2.2 MB, less than 21.43% of Rust online submissions for Bulls and Cows.
use std::collections::HashMap;

impl Solution {

    pub fn get_hint(secret: String, guess: String) -> String {

        pub fn check_bulls(secret_dictionary:&mut HashMap<char, i32>, guess_dictionary:&mut HashMap<char,i32>, secret: &String, guess: &String, bulls:&mut i32) -> (){
        for i in 0..secret.len(){
            if secret.as_bytes()[i] == guess.as_bytes()[i]{
                *bulls += 1;
                secret_dictionary.insert(secret.chars().nth(i).unwrap(), &secret_dictionary[&secret.chars().nth(i).unwrap()]-1);
                guess_dictionary.insert(guess.chars().nth(i).unwrap(), &guess_dictionary[&guess.chars().nth(i).unwrap()]-1 );
            }
        }
    }

        pub fn check_guess(secret_dictionary:&mut HashMap<char, i32>, guess_dictionary:&mut HashMap<char,i32>, secret: &String, guess: &String, cows:&mut i32) -> (){
        for i in secret.chars(){
            if secret_dictionary[&i] > 0 && guess_dictionary.contains_key(&i) && guess_dictionary[&i]>0{
                *cows += 1;
                secret_dictionary.insert(i, &secret_dictionary[&i]-1);
                guess_dictionary.insert(i, &guess_dictionary[&i]-1 );
            }
        }
    }

        let mut secret_hashmap:HashMap<char, i32> = HashMap::new();
        let mut guess_hashmap:HashMap<char, i32> = HashMap::new();
        let mut cows = 0;
        let mut bulls = 0;
        for i in secret.chars(){
            if secret_hashmap.contains_key(&i){
                secret_hashmap.insert(i, 1 + secret_hashmap[&i]);
            }
            else{
                secret_hashmap.insert(i,1 );
            }
        }
        for i in guess.chars(){
            if guess_hashmap.contains_key(&i){
                guess_hashmap.insert(i, 1 + guess_hashmap[&i]);
            }
            else{
                guess_hashmap.insert(i,1 );
            }
        }
        check_bulls(&mut secret_hashmap, &mut guess_hashmap, &secret, &guess, &mut bulls);
        check_guess(&mut secret_hashmap, &mut guess_hashmap, &secret, &guess, &mut cows);
        
        let result_string: String = format!("{}{}{}{}",&bulls.to_string().to_owned(),"A",&cows.to_string().to_owned(),"B" );
        result_string
    }
    
}